sympy/matrices/matrices.py

```diff
@@ -2478,10 +2478,14 @@ def do(x):
             if len(args) == 3 and isinstance(args[2], Callable):
                 op = args[2]
                 flat_list = []
-                for i in range(rows):
+                rows_range = cols_range = [cls._sympify(i) for i in range(max(rows, cols))]
+                if rows < cols:
+                    rows_range = rows_range[:rows]
+                elif cols < rows:
+                    cols_range = cols_range[:cols]
```

Comment:
I think that the performance improvement is not visible for the last changes, from my experiment,
Probably because of the nature that linear time complexity does not contribute significantly to the matrix.
But I'm keeping this because there was this specific part is faster for creating a list with a million elements.

Reply:
How about a *single flag* that controls only *output*:
`sympify_output` (which defaults to False).
```
ix = [_sympify(i) for i in range(max(rows, cols))]
rows_range = ix[:rows_range]
cols_range = ix[:cols_range]
for i in rows_range:
    flat.extend([f(i,j) for j in cols_range])
if sympify_output:
    flat = map(_sympify,flat)
```